### PR TITLE
fsdp with custom process groups

### PR DESCRIPTION
**Pull request by vchiley**
# What does this PR do?

This PR enables FSDP _auto_wrap to accept and propagate custom arguments, most notably, it enables the user to propagate dist process_groups.
Custom process_groups enable users to instantiate fsdp modules which shard / synchronize parameters over a subset of accelerators. This is useful for MoE Expert layers and TensorParallel layers.

# What issue(s) does this change relate to?

Mixture of Experts (MoE)
https://mosaicml.atlassian.net/browse/RESEARCH-351
https://mosaicml.atlassian.net/browse/CO-1716
https://mosaicml.atlassian.net/browse/CO-1715
https://mosaicml.atlassian.net/browse/CO-1714
https://mosaicml.atlassian.net/browse/CO-1712

TensorParallel (TP)
https://mosaicml.atlassian.net/browse/CO-1635
https://mosaicml.atlassian.net/browse/RESEARCH-442

In general this doesn't fix an issue, but enables a feature which enable TP and MoE models
https://github.com/mosaicml/examples/pull/180
also see: https://github.com/mosaicml/tutel/pull/1

btw I ran [this](https://github.com/mosaicml/examples/pull/180/files#diff-6db87bde9cfb73791e1be455afa625ad3dabc4ea723aa1eb56a10c5b83240914R213) "test" on 16 gpus with different process_group configurations to validate different configurations run. For the MoE setup with pg 'self' or 'set1', I do test if the different configurations result in the expert weights not being sync'd. I do not check if params / grads are sync'd in appropriately in the TP setting; since TP params are sharded, the params shouldn't be equal, so idk what to check, just the fact that they params are gathered, have the appropriate shape, and run might be test enough.

# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [x] Did you update any related docs and document your change?
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

Tests: this PR enables "super-users" to use advanced features and defaults to previous behavior. All former tests passing is test enough.

### CODE REVIEW

**Comment by dskhudia** [comment]
This also allows user to specify per module sharding strategy as well?

File: composer/trainer/mosaic_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    else:
+        raise ValueError(f'Unsure how to setup process_group={pg}')
+
+    ranks_per_subgroup_list = list(set(dist.all_gather_object(ranks)))
+    current_group, _subgroups = distributed.distributed_c10d.new_subgroups_by_enumeration(ranks_per_subgroup_list)
+    process_group_cache[pg_key] = current_group
+    return current_group
+
+
+def _pro_recursive_wrap(module: nn.Module,
```
Comment (dskhudia) [inline composer/trainer/mosaic_fsdp.py:86]: It's immediately not clear the pro is shorted version of process. Consider changing name to custom_recursive_wrap?

File: composer/trainer/mosaic_fsdp.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    criteria with :func:`auto_wrap`. Does not rely on _ConfigAutoWrap.
+
+    Args:
+        module (nn.Module):
+            module to recursively wrap
+        auto_wrap_policy (Callable):
+            A callable specifying a policy to recursively wrap layers with FSDP.
+        ignored_modules (Set[torch.nn.Module]): Modules to ignore when
+            wrapping.
+        ignored_params (Set[torch.nn.Parameter]): Parameters to ignore when
```
Comment (dskhudia) [inline composer/trainer/mosaic_fsdp.py:109]: Missing args process_group_cache and others here.

**Comment by vchiley** [comment]
https://github.com/mosaicml/composer/pull/2006#issuecomment-1445743125
Yes this enables the user to pass custom args to fsdp for every fsdp wrapped module; we include some of the mosaicml tooling for easily setting the configs
cc @dskhudia

File: composer/trainer/mosaic_fsdp.py [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+    """Convert common string representations of dtypes to torch dtypes."""
+    dtype = dtype.value if isinstance(dtype, Precision) else dtype
+    if dtype in ['float32', 'torch.float32', 'fp32']:
+        return torch.float32
+    elif dtype in ['float16', 'torch.float16', 'half', 'fp16', 'amp', 'amp_fp16']:
+        return torch.float16
+    elif dtype in ['bfloat16', 'bfloat', 'torch.bfloat16', 'bf16', 'amp_bf16']:
+        return torch.bfloat16
+    elif dtype in ['amp_fp8']:
+        # We use torch.bfloat16 by default for amp_fp8 as there is no
```
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:58]: Do we want to throw a warning here for now? + a TODO when fp8 is supported in PyTorch?

Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:58]: This was [copy pasta from dist_strategy.py](https://github.com/mosaicml/composer/pull/2006/files/1e9ebe27db16f25c3528dbd8d4c53c6652f49757#diff-dd51f7e9f15c34f3f9e4895d122b92bf4e8e713cdf582a4af5e5e2eef628c0d1L126), but yes adding a warning sounds like a good idea. will do.

File: composer/trainer/mosaic_fsdp.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+        warnings.warn(f"Converting process_group='node' to process_group='{pg}'")
+    elif pg == 'local_rank_across_nodes':
+        pg = f'mod{world_size}'
+        warnings.warn(f"Converting process_group='local_rank_across_nodes' to process_group='{pg}'")
+
+    # Handle str and Union[List[int], Tuple[int]] process_group cases
+    if isinstance(pg, str) and pg.startswith('set'):
+        k = int(pg.strip('set'))
+        world_size = dist.get_world_size()
+        if world_size % k:
```
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:138]: Don't we need an equality check `== 0` here?

File: composer/trainer/mosaic_fsdp.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+            raise RuntimeError(f'{world_size} must be divisible by set size ({k})')
+        start = dist.get_global_rank() // k * k
+        ranks = tuple(range(start, start + k))
+    elif isinstance(pg, str) and pg.startswith('mod'):
+        k = int(pg.strip('mod'))
+        world_size = dist.get_world_size()
+        if world_size % k:
```
Comment (bcui19) [inline composer/trainer/mosaic_fsdp.py:145]: same as above with set?

File: composer/trainer/mosaic_fsdp.py (hunk 4 above)
Comment (vchiley) [reply composer/trainer/mosaic_fsdp.py:141]: if `world_size` is divisible by `k` then `world_size % k == 0` and we do not need to raise the error.
with an `== 0` check you'd also need to add a not
```
if not world_size % k == 0:
```

right?

**Comment by bcui19** [comment]
LGTM, will approve when we add docs

**Review by mvpatel2000 — CHANGES_REQUESTED** [review]
I'm going to hold this PR until we get multiple approval here just because it looks scary and might break release, so we should be extra careful. Will lift once we have sign-offs from all parties. This is basically in lieu of codeowners for this part of the repo, which isn't configured.

Requiring approvals from @bcui19 @dakinggg and myself.

File: composer/trainer/mosaic_fsdp.py (hunk 4 above)
Comment (dakinggg) [reply composer/trainer/mosaic_fsdp.py:141]: `world_size % k != 0` would be a bit clearer I think, just so I don't have to think about which case is which

File: composer/trainer/mosaic_fsdp.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        ranks = tuple(range(dist.get_global_rank() % k, world_size, k))
+    elif isinstance(pg, (list, tuple)):
+        ranks = tuple(pg)
+    else:
+        raise ValueError(f'Unsure how to setup process_group={pg}')
+
+    if process_group_cache is not None and ranks in process_group_cache:
+        warnings.warn(
+            f'On rank={dist.get_global_rank()} using cached progress group with {ranks=}. ' +\
+            'Instantiate new process group if this is what was intended.'
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:159]: I don't understand the second sentence here. can you clarify?

File: composer/trainer/mosaic_fsdp.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+class MosaicFullyShardedDataParallel(FullyShardedDataParallel):
+    """Updates FSDP's _auto_wrap to enalbe module_kwargs."""
+
+    def _auto_wrap(
+        self,
+        auto_wrap_kwargs: Dict[str, Any],
+        fsdp_kwargs: Dict[str, Any],
+    ) -> None:
+        """Updates _auto_wrap to enalbe module_kwargs.
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:282]:
```suggestion
        """Updates _auto_wrap to enable module_kwargs.
```

File: composer/trainer/mosaic_fsdp.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+                           auto_wrap_policy: Callable,
+                           wrapper_cls: Callable,
+                           ignored_modules: Set[nn.Module],
+                           ignored_params: Set[nn.Parameter],
+                           process_group_cache: Dict[Tuple[int], Any],
+                           only_wrap_children: bool = False,
+                           **kwargs: Any) -> Tuple[nn.Module, int]:
+    """Updates FSDPs _recursive_wrap to enable module_kwargs and custom process_group cache.
+
+    modified version of
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:186]: could you call out what specifically is modified?

File: composer/trainer/mosaic_fsdp.py [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    """Updates FSDP's _auto_wrap to enalbe module_kwargs."""
+
+    def _auto_wrap(
+        self,
+        auto_wrap_kwargs: Dict[str, Any],
+        fsdp_kwargs: Dict[str, Any],
+    ) -> None:
+        """Updates _auto_wrap to enalbe module_kwargs.
+
+        modified version of
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:284]: could you call out what specifically is modified?

File: composer/trainer/mosaic_fsdp.py [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+from torch.distributed.fsdp.wrap import _or_policy, _wrap, _wrap_batchnorm_individually
+
+from composer.core import Precision
+from composer.utils import dist
+
+__all__ = [
+    'sharding_map',
+    'backward_prefetch_map',
+    'get_torch_dtype',
+    '_get_mixed_precision',
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:29]: these aren't really private functions since they're being exported, so maybe remove the `_` and add docstrings?

**Review by dakinggg — COMMENTED** [review]
I think this looks good to me, left a few minor comments. Could you also test that a mosaicgpt-125m runs and loss curve looks reasonable? (i.e. make sure this PR doesn't break stuff unrelated to this PR)

**Comment by vchiley** [comment]
@bcui19 added docs
@dakinggg made suggested updts, suggested runs from diff branches [here](https://wandb.ai/mosaic-ml/composer_mfsdp_test) (using [tutel_moe branch](https://github.com/vchiley/examples/tree/tutel_moe) in examples repo since it points to this branch)

File: docs/source/notes/distributed_training.rst [hunk 11]
```diff
@@ -306,6 +306,61 @@ In `gpt.py <https://github.com/mosaicml/examples/blob/6972fe3000d5a5480d8757ff71
 A very similar auto wrap policy is provided for activation checkpointing, with analogous rule #1 that looks for :code:`module._activation_checkpointing = True | False` and rule #2 that looks for :code:`def activation_checkpointing_fn(module: torch.nn.Module) -> bool`.
 
 
+**Experimental:** Composer enables the users to specify custom FSDP args for all wrapped modules. This is enabled by returning a dictionary of args instead of returning a bool.
```
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:309]:
```suggestion
**Experimental:** Composer enables users to specify custom FSDP args for all wrapped modules. This is enabled by returning a dictionary of args instead of returning a bool.
```

File: docs/source/notes/distributed_training.rst [hunk 12]
```diff
@@ -306,6 +306,61 @@ In `gpt.py <https://github.com/mosaicml/examples/blob/6972fe3000d5a5480d8757ff71
...
+    class Block(nn.Module):
+        ...
+
+    class BlockRequiringCustomArgs(nn.Module):
+        ...
+
+    class Model(nn.Module):
+        def __init__(self, n_layers):
+            super().__init__()
+            self.blocks = nn.ModuleList([
```
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:324]: maybe this model should actually contain some `BlockRequiringCustomArgs`?

File: docs/source/notes/distributed_training.rst [hunk 13]
```diff
@@ -306,6 +306,61 @@ In `gpt.py <https://github.com/mosaicml/examples/blob/6972fe3000d5a5480d8757ff71
...
+
+        # Activation Checkpointing Function
+        def activation_checkpointing_fn(self, module):
+            return isinstance(module, Block)
+
+While the user can instantiate and pass in process groups, Composer enables process groups to be specified using the following options:
+
+1. :code:`'self'`: the degenerate case where all process groups only operate within their current rank (:code:`'self'` == :code:`'set1'`). This is useful when you do not want a layer to be synchonized across accelerators.
+
+2. :code:`'node'`: instantiates process groups which opereate within a node (:code:`'mode'` == :code:`f'set{local_world_size}'`). This is useful for Expert Layers in MoE models.
```
Comment (dakinggg) [inline docs/source/notes/distributed_training.rst:355]:
```suggestion
2. :code:`'node'`: instantiates process groups which opereate within a node (:code:`'node'` == :code:`f'set{local_world_size}'`). This is useful for Expert Layers in MoE models.
```

File: composer/trainer/mosaic_fsdp.py [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+
+            # Leaf node or final wrapping of the remainder both happen here.
+            return _wrap(module, wrapper_cls, **final_kwargs), num_params
+        else:
+            return module, total_wrapped_params
+    return module, 0
+
+
+class MosaicFullyShardedDataParallel(FullyShardedDataParallel):
+    """Updates FSDP's _auto_wrap to enalbe module_kwargs."""
```
Comment (dakinggg) [inline composer/trainer/mosaic_fsdp.py:281]:
```suggestion
    """Updates FSDP's _auto_wrap to enable module_kwargs."""
```

**Review by dakinggg — APPROVED** [review]
LGTM

File: docs/source/notes/distributed_training.rst (hunk 12 above)
Comment (vchiley) [reply docs/source/notes/distributed_training.rst:324]: valid point

**Review by bcui19 — APPROVED** [review]
LGTM!

**Review by mvpatel2000 — APPROVED** [review]
LGTM.
